static/app/utils/useAsync.tsx

```diff
@@ -0,0 +1,351 @@
+import React, {useEffect, useState} from 'react';
+import * as Sentry from '@sentry/react';
+
+import {ResponseMeta} from 'sentry/api';
+import LoadingError from 'sentry/components/loadingError';
+import LoadingIndicator from 'sentry/components/loadingIndicator';
+import {t} from 'sentry/locale';
+import {metric} from 'sentry/utils/analytics';
+import getRouteStringFromRoutes from 'sentry/utils/getRouteStringFromRoutes';
+import useApi from 'sentry/utils/useApi';
+import useLocation from 'sentry/utils/useLocation';
+import useParams from 'sentry/utils/useParams';
+import useRoutes from 'sentry/utils/useRoutes';
+import PermissionDenied from 'sentry/views/permissionDenied';
+import RouteError from 'sentry/views/routeError';
+
+type State = {
+  data: {[key: string]: any};
+
+  errors: {[key: string]: ResponseMeta};
+  hasError: boolean;
+  /**
+   * This is state for when fetching data from API
+   */
+  isLoading: boolean;
+  isReloading: boolean;
+  remainingRequests?: number;
+
+  /**
+   * Indicates that Team results (from API) are paginated and there are more
+   * Teams that are not in the initial response.
+   */
+};
+
+type Result = {
+  renderComponent: (_child: React.ReactNode) => React.ReactNode;
+} & State;
+
+type Options = {
+  disableErrorReport: boolean;
+  reloadOnVisible: boolean;
+  shouldReload: boolean;
+  shouldReloadOnVisible: boolean;
+  shouldRenderBadRequests: boolean;
+};
+
+type Props = {
```

Comment:
`Props` isn't quite the right type name here, it's really more like `HookOptions`. I would say `Options`, but that's already taken